Add support for cmse_nonsecure_entry attribute

hug-dev opened the pull request:
This pull request adds the `cmse_nonsecure_entry` attribute under an unstable feature.

I was not sure if it was fine for me to send directly the pull-request or if I should submit a RFC first. I was told on Zulip that it was fine to do so but please close it if I need first submit a RFC or follow another process instead.

The `cmse_nonsecure_entry` attribute is a LLVM attribute that will be available in LLVM 11. I plan to rebase on the [upgrade PR](https://github.com/rust-lang/rust/pull/73526) once merged to make this one compile.

This attribute modifies code generation of the function as explained [here](https://developer.arm.com/documentation/ecm0359818/latest/) to make it work with the TrustZone-M hardware feature. This feature is only available on `thumbv8m` targets so I created an error for that if one tries to use this attribute for another target.

I added this attribute in Rust as any other LLVM attribute are added but since this one is target-dependent I am not sure if it was the best thing to do. Please indicate me if you think of other ways, like isolating target-dependent attributes together.

----------------

Tracking issue: https://github.com/rust-lang/rust/issues/75835

jonas-schievink reviewed (commented):
Compiler changes look basically good to me. The whole PR is so small that I think we can add this feature without an RFC/MCP/FCP/whatever, but stabilizing the feature will require more care.

I think a few small things need to be done before this can land:

* The PR needs a rebase to fix the merge conflicts.
* The new feature needs to be added to and documented in the unstable book (`src/doc/unstable-book`).
* A UI test should be added that ensures that the error you added is emitted.
* A tracking issue needs to be opened.

It would also be good to figure out more about plans for CMSE support in Rust. For example, is this the only thing that needs compiler support to allow writing privileged Rust apps that expose an interface for unprivileged code, or is more needed? If more language features are needed, it would be a good idea to coordinate with the language team to figure out the bigger picture before implementing more things.

hug-dev commented on src/doc/unstable-book/src/language-features/cmse-nonsecure-entry.md:
```diff
@@ -0,0 +1,75 @@
+With this attribute, the compiler will do the following:
+* add a special symbol on the function which is the `__acle_se_` prefix and the
+  standard function name
+* constrain the number of parameters to avoid using the Non-Secure stack
```
I think I agree with that, an LLVM error will be a bit more cryptic for the user. I am going to look into how and were I can do that. Basically we want to emit an error if the function is going to pass any of its parameter using the stack as opposed to using only registers.

jonas-schievink replied:
I think it would also be fine to do this as follow-up work if you want. Just add it to the blockers in the tracking issue since it's important to do before stabilization.

hug-dev replied:
The LLVM check was accidentally omitted and [is being reviewed](https://reviews.llvm.org/D86478) now.
I am happy for this to be done as a follow-up work but I think the LLVM patch should be there (once merged). If not, entry functions declaring more than 32 bytes of arguments would at best provoke a segmentation fault but at worse expose secure data...

Or is that fine for unstable features not to be safe/secure if that is documented somewhere?

jonas-schievink replied:
Ah, right, yeah. If this can cause segfaults we should wait for that fix.

hug-dev replied:
Then I guess I can follow [this guide](https://rustc-dev-guide.rust-lang.org/backend/updating-llvm.html#bugfix-updates) to cherry-pick the bugfix into the LLVM fork Rust uses.

hug-dev replied:
Regarding tests, would it work if I add a UI test with a `// compile-flags: --target thumbv8m.main-none-eabi` option ([similar to this](https://github.com/rust-lang/rust/blob/master/src/test/ui/issues/issue-49851/compiler-builtins-error.rs)) and that I check that the LLVM error is generated when declaring an entry function with too many parameters?

jonas-schievink replied:
In addition to that, you'll need to let the test runner know that the test only works on the `thumbv8m.main-none-eabi` target so it is skipped when that target isn't being built. I think that can be done with this directive:

`// only-thumbv8m.main-none-eabi`

Could be wrong though.

hug-dev replied:
The LLVM bug fix is now merged! I made [a PR](https://github.com/rust-lang/llvm-project/pull/76) to cherry-pick it

bors commented:
:umbrella: The latest upstream changes (presumably #74862) made this pull request unmergeable. Please resolve the merge conflicts.

nihalpasham commented:
Thought this may be useful to folks interested in the PR. I published a small medium post, summarizing TZ for ARMv8-M (as a future reference to myself).  

https://bit.ly/2Z3DWd8

Commit:
Add support for cmse_nonsecure_entry attribute

This patch adds support for the LLVM cmse_nonsecure_entry attribute.
This is a target-dependent attribute that only has sense for the
thumbv8m Rust targets.
You can find more information about this attribute here:
https://developer.arm.com/documentation/ecm0359818/latest/

Signed-off-by: Hugues de Valon <hugues.devalon@arm.com>

Commit:
Update LLVM and add Unsupported diagnostic

Secure entry functions do not support if arguments are passed on the
stack. An "unsupported" diagnostic will be emitted by LLVM if that is
the case.
This commits adds support in Rust for that diagnostic so that an error
will be output if that is the case!

Signed-off-by: Hugues de Valon <hugues.devalon@arm.com>

hug-dev commented:
Hey @jonas-schievink 👋 !

Sorry for the force-push but since when rebasing I realised that everything changed under the feet of this PR, I gathered I might as well 😅 

I added a first commit that updates the LLVM submodule to the one that includes the bug fix. The bug fix is there but when creating an entry function which requires arguments being passed on stack, it throws an "unsupported" [diagnostic](https://github.com/llvm/llvm-project/blob/7fcad5583a12026ce19afe487681753ac633064a/llvm/lib/Target/ARM/ARMISelLowering.cpp#L4447) which was not supported in Rust ABI. I added support for it in the first commit as well.

Also added a test for that case, the error message:
```
error: <unknown>:0:0: in function entry_function i32 (i32, i32, i32, i32, i32): secure entry function requires arguments on stack
```
weirdly lacks location information but at least you have the function name so I thought it was enough for now.

I don't know how to pass that UI test locally though so I am sure it will fail on the CI 😢

jonas-schievink commented on src/test/ui/cmse-nonsecure-entry/params-on-stack.rs:
```diff
@@ -0,0 +1,11 @@
+// gate-test-cmse_nonsecure_entry
```
A feature gate test tests that an unstable feature is not usable without a `#![feature]` attribute, so you need to add a test that does so and move this comment there

```suggestion
```

hug-dev replied:
Added the gate-test and fixed the wrong header in other ones

jonas-schievink commented on src/test/ui/cmse-nonsecure-entry/params-on-stack.rs:
```diff
@@ -0,0 +1,11 @@
+
+#[no_mangle]
+#[cmse_nonsecure_entry]
+pub extern "C" fn entry_function(a: u32, b: u32, c: u32, d: u32, e: u32) -> u32 {
```
We should also have a test that does the same thing, except with just 4 arguments (as a build-pass test).

hug-dev replied:
Added such test!

jonas-schievink commented on src/test/ui/cmse-nonsecure-entry/trustzone-only.rs:
```diff
@@ -0,0 +1,11 @@
+// gate-test-cmse_nonsecure_entry
```
```suggestion
```

jonas-schievink commented on src/test/ui/cmse-nonsecure-entry/params-on-stack.stderr:
```diff
@@ -0,0 +1,5 @@
+error: <unknown>:0:0: in function entry_function i32 (i32, i32, i32, i32, i32): secure entry function requires arguments on stack
```
Yeah this seems fine for an initial version of this feature, but the error should be moved into rustc eventually (we know some function ABI details already, and this seems fairly easy to check). We can add a note about that to the tracking issue.

hug-dev replied:
Added to the tracking issue.

jonas-schievink reviewed (commented):
Nice! Left some more comments.

jonas-schievink commented on src/doc/unstable-book/src/language-features/cmse-nonsecure-entry.md:
```diff
@@ -0,0 +1,75 @@
+5.4](https://developer.arm.com/documentation/ecm0359818/latest/) for details).
+With this attribute, the compiler will do the following:
+* add a special symbol on the function which is the `__acle_se_` prefix and the
+  standard function name
```
Ah, I think we should definitely forbid using this attribute with non-`"C"` ABI functions then. Since the Rust ABI is unstable, it can pass arguments either directly in registers, or in memory, and that would be a problem here.

hug-dev replied:
Yes that makes sense, I am looking if there is a way to check what the ABI of the function compiled is.

hug-dev replied:
It was easier than anticipated 😃 Added a new error code for that.

Commit:
Add more tests and check for ABI

Signed-off-by: Hugues de Valon <hugues.devalon@arm.com>

jonas-schievink commented:
@bors r+

bors commented:
:pushpin: Commit 2588287deffd92c4ca0b5b39f92b9cca894a3983 has been approved by `jonas-schievink`

<!-- @bors r=jonas-schievink 2588287deffd92c4ca0b5b39f92b9cca894a3983 -->
<!-- homu: {"type":"Approved","sha":"2588287deffd92c4ca0b5b39f92b9cca894a3983","approver":"jonas-schievink"} -->

bors commented:
:hourglass: Testing commit 2588287deffd92c4ca0b5b39f92b9cca894a3983 with merge 867bd42c38bac3870388765f65386b18cf1d9c5f...
<!-- homu: {"type":"BuildStarted","head_sha":"2588287deffd92c4ca0b5b39f92b9cca894a3983","merge_sha":"867bd42c38bac3870388765f65386b18cf1d9c5f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1190032753), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36938)
Approved by: jonas-schievink
Pushing 867bd42c38bac3870388765f65386b18cf1d9c5f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jonas-schievink","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36938","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1190032753"},"merge_sha":"867bd42c38bac3870388765f65386b18cf1d9c5f"} -->

nihalpasham commented:
> There is also another attribute to be implemented, `cmse_nonsecure_call` which has the specificity to be a call site attribute. You can use it from the Secure code to be able to call a Non-Secure function pointer like:
> 
> ```rust
> fn non_secure_callback(callback: fn(i32) -> i32) {
>     let val = (#[cmse_nonsecure_call] callback)(42);
> }
> ```
> 
> This would need a new Rust feature to be able to add call site attributes. This is also gated by the `stmt_expr_attributes` feature which is not stable yet. I have made a local version that works but only for this specific attribute. I can make a RFC/MCP/FCP to propose a design for this feature and have `cmse_nonsecure_call` be the first example?
> 
> The other things needed to build a TrustZone-M application are mostly in `cortex-m` already: abstraction over the `TT*` instructions and Secure Attribution Unit (SAU) support. I am planning to make a blog post to summarize all those features showing an example.

Hi @hug-dev - so, I've been experimenting with this PR in nightly over the weekend and firstly thank you so much for the work 👍. Just a quick question or rather clarification - From the above comments, I understand we still need to wait for support to be added for `call-site attributes` in Rust before we can build a fully functioning demo i.e. TrustZone-M hands-off control to non-secure code from the secure world upon initial bootup and we require the call-site attr for the first callback from secure code. - right

If my understanding is correct - would you happen to know if anyone in the Rust community/team is working on call-site attributes support? or if we can put in a PR for this without duplication.

hug-dev commented:
Thank you 😄 I am glad this is useful.

> TrustZone-M hands-off control to non-secure code from the secure world upon initial bootup and we require the call-site attr for the first callback from secure code.

You are right that you could use the `cmse_nonsecure_call` for the first switch Secure -> Non-Secure but you could also do it *manually*!
In a local example I did the following:

```rust
unsafe {
    // The beginning of the flash memory for the Non-Secure application
    let ns_vector_table = 0x00200000u32 as *const u32;
    // Set the last bit to zero to show a transition from Secure to Non-Secure
    let ns_reset_vector = *(0x00200004u32 as *const u32) & !1u32;
    // Switch the stack pointer to the Main Stack Pointer from Non-Secure world
    cortex_m::register::msp::write_ns(*ns_vector_table);
    // Branch to the Non-Secure reset vector
    cortex_m::asm::bx_ns(ns_reset_vector);
}
```

I added `write_ns` and `bx_ns` in `cortex-m` just for that 🤓!
Note that this is not really secure as it does not clear the Secure registers before jumping so information could be leaked there. Instead, if `cmse_nonsecure_call` was there, it could be something like:

```rust
unsafe {
    // The beginning of the flash memory for the Non-Secure application
    let ns_vector_table = 0x00200000u32 as *const u32;
    cortex_m::register::msp::write_ns(*ns_vector_table);

    let ns_entry: extern "C" fn() = unsafe { std::mem::transmute(*(0x00200004u32 as *const u32)) };
    (#[cmse_nonsecure_call] ns_entry)()
}
```

I don't think anyone is currently working on call site attributes, and I was meant to start a "design proposal" of how to implement this attribute. There was also a suggestion to encode it as a different ABI. If you are definitely interested in this one, I can get to it sooner than later!

nihalpasham commented:
> I don't think anyone is currently working on call site attributes, and I was meant to start a "design proposal" of how to implement this attribute. There was also a suggestion to encode it as a different ABI. If you are definitely interested in this one, I can get to it sooner than later!

That would be great.... definitely `very` interested in getting this to work (so as to avoid `relying on C` for TrustZone setup and usage). I'm working on a project of mine where I'm evaluating different kinds of embedded -Hardware Accelerators and Root(s) of Trust (examples below) and one of my requirements is a `secure runtime environment` for hosting the driver used to communicate with an HRT. 
- https://github.com/nihalpasham/nrf-secureboot-test
- https://github.com/nihalpasham/Rusty_CryptoAuthLib

So, yep I'm most interested and would also love to help in any way I can. {although full disclosure - my knowledge on compilers is `it magically compiles` 😅 }

hug-dev commented:
Nice! Well I can start by making a proposal on Zulip that I will post here when ready. Then you can help reviewing it 😃

nihalpasham commented:
@hug-dev Hi, just checking in to see if there are any new updates.

hug-dev commented:
Hey @nihalpasham!
Arg sorry, I was working on other stuff and it completely slipped my mind. I am soon going to holidays for the Christmas break but I will make a note to myself to resurrect my notes and work on this and start a discussion around this issue in the beginning of Jannaury. I hope that's ok with you!

nihalpasham commented:
Oh, that's alright. I kind of lost track of this too. 

PS - Wish you a merry Christmas and happy holidays!

nihalpasham commented:
Hey @hug-dev,  Just thought I'd check-in to see if you're back.

hug-dev commented:
Hello! I am 😄 Ok, I will take some time this week or the next to think about this again and start a discussion on Zulip 👍

hug-dev commented:
Discussion about how to add the remaining `cmse_nonsecure_call` feature is happimh [here on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Support.20for.20callsite.20attributes/near/223054928).
